src/lib/mongoBackend/MongoGlobal.cpp

```diff
@@ -2045,3 +2045,113 @@ void releaseTriggeredSubscriptions(std::map<std::string, TriggeredSubscription*>
   }
   subs.clear();
 }
+
+/* ****************************************************************************
+*
+* fillContextProviders -
+*
+* This functions is very similar the one with the same name in mongoQueryContext.cpp
+* but acting on a single CER instead that on a complete vector of them.
+*
+* Looks in the CER passed as argument, searching for a suitable CPr in the CRR
+* vector passed as argument. If a suitable CPr is found, it is set in the CER (and the 'found' field
+* is changed to true)
+*
+*/
+void fillContextProviders(ContextElementResponse* cer, ContextRegistrationResponseVector& crrV)
+{
+
+  for (unsigned int ix = 0; ix < cer->contextElement.contextAttributeVector.size(); ++ix)
+  {
+    ContextAttribute* ca = cer->contextElement.contextAttributeVector.get(ix);
+    if (ca->found)
+    {
+      continue;
+    }
+    /* Search for some CPr in crrV */
+    std::string perEntPa;
+    std::string perAttrPa;
+    searchCprForAttribute(cer->contextElement.entityId, ca->name, crrV, &perEntPa, &perAttrPa);
+
+    /* Looking results after crrV processing */
+    ca->providingApplication = perAttrPa == ""? perEntPa : perAttrPa;
+    ca->found = (ca->providingApplication != "");
+  }
+
+}
+
+/* ****************************************************************************
+*
+* someContextElementNotFound -
+*
+* This functions is very similar the one with the same name in mongoQueryContext.cpp
+* but acting on a single CER instead that on a complete vector of them.
+*
+* Returns true if some attribute with 'found' set to 'false' is found in the CER passed
+* as argument
+*
+*/
+bool someContextElementNotFound(ContextElementResponse& cer)
+{
+
+  for (unsigned int ix = 0; ix < cer.contextElement.contextAttributeVector.size(); ++ix)
+  {
+    if (!cer.contextElement.contextAttributeVector[ix]->found)
+    {
+      return true;
+    }
+  }
+  return false;
+}
+
+
+/* ****************************************************************************
+*
+* searchCprForAttribute -
+*
+* Search the CPr, given the entity/attribute as argument. Actually, two CPrs can be returned
+* the "general" one at entity level or the "specific" one at attribute level
+*
+*/
+void searchCprForAttribute(EntityId&                          en,
+                           const std::string&                 attrName,
+                           ContextRegistrationResponseVector& crrV,
+                           std::string*                       perEntPa,
+                           std::string*                       perAttrPa)
+{
+  *perEntPa  = "";
+  *perAttrPa = "";
+  for (unsigned int ix = 0; ix < crrV.size(); ++ix)
+  {
+    ContextRegistrationResponse* crr = crrV.get(ix);
+    /* Is there a matching entity in the CRR? */
+    for (unsigned jx = 0; jx < crr->contextRegistration.entityIdVector.size(); ++jx)
+    {
+      EntityId* regEn = crr->contextRegistration.entityIdVector.get(jx);
+      /* No match (keep searching in other CRR) */
```

Comment:
If you put this comment inside the "if-block", it will be more clear:

    if (regEn->id != en.id || (regEn->type != en.type && regEn->type != ""))
    {
      // No match (keep searching in other CRR)
      continue
    }